rgb.man
.\" Copyright 1994, 2008 Sun Microsystems, Inc.  All rights reserved.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice (including the next
.\" paragraph) shall be included in all copies or substantial portions of the
.\" Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.TH rgb __appmansuffix__ __xorgversion__
.IX "rgb" "" "\f3rgb\fP(1) \(em build colorname database" ""
.IX "color" "build" "color" "building the colorname database \(em \f3rgb\fP(1)" ""
.SH NAME
rgb \- build the color name database
.SH SYNOPSIS
.B rgb
[
.I dbname
]
.SH DESCRIPTION
.B rgb
reads from standard input lines of the form:
.PP
.nf
.I
	red\ \ green\ \ blue\ \ \ \ \ \ name
.fi
.PP
where
.I red
/
.I green
/
.I blue
are decimal values between the range 0 to 255, and
.I name
is a description of the color.
.B rgb
then builds a color name database in
.B __RGB_DB_TYPE__
format.  The color name database provides a mapping between ASCII color
names and RGB color values.  It is useful for increasing the
portability of color programs.  The input source for the database is
in \fI__RGB_DB__.txt\fP.  \fIrgb.txt\fP
is compiled into the __RGB_DB_TYPE__ files
.IR "__RGB_DB_FILES__" .
When the server first starts up, it consults the contents of these
files to build an internal representation of their contents.  This
internal representation is consulted to map color names to color
values.
.PP
.B rgb
uses the default color name database of \fI__RGB_DB__.txt\fP.
.SH OPTIONS
.IP "\fIdbname" 1i
Color name database.
.SH EXAMPLES
example% rgb rgb < __RGB_DB__.txt
.PP
example% cat __RGB_DB__.txt
.SH FILES
.PD 0
.TP 30
.B __RGB_DB__.txt
color name database source.
Maps color names to RGB color values.
.TP
.B __RGB_DB_DIR__/__RGB_DB_FILES__
__RGB_DB_TYPE__ files containing color name to RGB mapping.
.PD
.SH SEE ALSO
.BR cat (1),
.BR __RGB_DB_TYPE__ (3)


ChangeLog
2006-04-28  Jeremy C. Reed  <reed@reedmedia.net>

	reviewed by: Adam Jackson

	* configure.ac:
	pkg-config check for xproto instead of x11.

2005-12-14  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for final X11R7 release candidate.

2005-12-06  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* Makefile.am:
	Change *man_SOURCES ==> *man_PRE to fix autotools warnings.

2005-12-03  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC3 release.

2005-12-01  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Change the default rgb dir to be the same path as the X server.

2005-11-23  Alan Coopersmith  <alan.coopersmith@sun.com>

	* configure.ac: 
	Fix --with-rgb-dir help string to give the right flag name.

2005-11-19  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* Makefile.am:
	* configure.ac:
	Update dependencies to work with separate build roots.

2005-11-09  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC2 release.

2005-10-18  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for RC1 release.



rgb.c
/* $Xorg: rgb.c,v 1.4 2001/02/09 02:05:35 xorgcvs Exp $ */
/* $XdotOrg: $ */
/*

Copyright 1985, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from The Open Group.

*/
/* $XFree86: xc/programs/rgb/rgb.c,v 3.8tsi Exp $ */


/* reads from standard input lines of the form:
	red green blue name
   where red/green/blue are decimal values, and inserts them in a database.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include DBM_HEADER
#ifndef NDBM
#define dbm_open(name,flags,mode) (!dbminit(name))
#define dbm_store(db,key,content,flags) (store(key,content))
#define dbm_close(db) dbmclose()
#endif

#undef NULL
#include <stdio.h>
#include <stdlib.h>
#include <X11/Xos.h>
#include "rgb.h"
#include <ctype.h>

#include <errno.h>

static char *ProgramName;

int
main(int argc, char **argv)
{
    char *dbname;
    char line[512];
    int red, green, blue;
    RGB rgb;
    datum key, content;
    char name[512];
    int items;
    int lineno;
    int i, n;
    int fd;
#ifdef NDBM
    DBM *rgb_dbm;
#else
    int rgb_dbm;
#endif

    ProgramName = argv[0];

    if (argc == 2)
	dbname = argv[1];
    else
	dbname = RGB_DB;

    strcpy (name, dbname);
    strcat (name, ".dir");
    fd = open (name, O_WRONLY|O_CREAT, 0666);
    if (fd < 0) {
	fprintf (stderr, 
		 "%s:  unable to create dbm file \"%s\" (error %d, %s)\n",
		 ProgramName, name, errno, strerror(errno));
	exit (1);
    }
    (void) close (fd);

    strcpy (name, dbname);
    strcat (name, ".pag");
    fd = open (name, O_WRONLY|O_CREAT, 0666);
    if (fd < 0) {
	fprintf (stderr, 
		 "%s:  unable to create dbm file \"%s\" (error %d, %s)\n",
		 ProgramName, name, errno, strerror(errno));
	exit (1);
    }
    (void) close (fd);

    rgb_dbm = dbm_open (dbname, O_RDWR|O_CREAT, 0666);
    if (!rgb_dbm) {
	fprintf (stderr,
		 "%s:  unable to open dbm database \"%s\" (error %d, %s)\n",
		 ProgramName, dbname, errno, strerror(errno));
	exit (1);
    }

    key.dptr = name;
    content.dptr = (char *) &rgb;
    content.dsize = sizeof (rgb);
    lineno = 0;
    while (fgets (line, sizeof (line), stdin)) {
	lineno++;
	if (line[0] == '!')
	    continue;
	items = sscanf (line, "%d %d %d %[^\n]\n", &red, &green, &blue, name);
	if (items != 4) {
	    fprintf (stderr, "syntax error on line %d\n", lineno);
	    fflush (stderr);
	    continue;
	}
	if (red < 0 || red > 0xff ||
	    green < 0 || green > 0xff ||
	    blue < 0 || blue > 0xff) {
	    fprintf (stderr, "value for %s out of range\n", name);
	    fflush (stderr);
	    continue;
	}
	n = strlen (name);
	for (i = 0; i < n; i++) {
	    if (isupper (name[i]))
		name[i] = tolower (name[i]);
	}
	key.dsize = n;
	rgb.red = (red * 65535) / 255;
	rgb.green = (green * 65535) / 255;
	rgb.blue = (blue * 65535) / 255;
	if (dbm_store (rgb_dbm, key, content, DBM_REPLACE)) {
	    fprintf (stderr, "%s:  store of entry \"%s\" failed\n",
		     ProgramName, name);
	    fflush (stderr);
	}
    }

    dbm_close(rgb_dbm);

    exit(0);
}
